# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/platform/unix/font.rs
# Commit: 793e9dfa4b9f
# Full Hash: 793e9dfa4b9fa40137cdf2a5d34c126d6cc9840d
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2020-02-29 04:04:56
# Regressor Bug: 1616995
# File Overlap Count: 1
# Description:
#   Bug 1616995 - patch 2 - Support vertical skew for upright-vertical fonts in webrender. r=lsalzman
#   
#   Depends on D63893
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D64746
# ==============================================================================

diff -r 88d3197f16c2 -r 793e9dfa4b9f gfx/wr/webrender/src/platform/unix/font.rs
--- a/gfx/wr/webrender/src/platform/unix/font.rs	Fri Feb 28 14:55:39 2020 +0000
+++ b/gfx/wr/webrender/src/platform/unix/font.rs	Fri Feb 28 15:36:19 2020 +0000
@@ -12,7 +12,7 @@
 use freetype::freetype::{FT_GlyphSlot, FT_LcdFilter, FT_New_Face, FT_New_Memory_Face};
 use freetype::freetype::{FT_Init_FreeType, FT_Load_Glyph, FT_Render_Glyph};
 use freetype::freetype::{FT_Library, FT_Outline_Get_CBox, FT_Set_Char_Size, FT_Select_Size};
-use freetype::freetype::{FT_Fixed, FT_Matrix, FT_Set_Transform, FT_String, FT_ULong};
+use freetype::freetype::{FT_Fixed, FT_Matrix, FT_Set_Transform, FT_String, FT_ULong, FT_Vector};
 use freetype::freetype::{FT_Err_Unimplemented_Feature, FT_MulFix, FT_Outline_Embolden};
 use freetype::freetype::{FT_LOAD_COLOR, FT_LOAD_DEFAULT, FT_LOAD_FORCE_AUTOHINT};
 use freetype::freetype::{FT_LOAD_IGNORE_GLOBAL_ADVANCE_WIDTH, FT_LOAD_NO_AUTOHINT};
@@ -224,7 +224,7 @@
 // a given FontContext so it is safe to move the latter between threads.
 unsafe impl Send for FontContext {}
 
-fn get_skew_bounds(bottom: i32, top: i32, skew_factor: f32) -> (f32, f32) {
+fn get_skew_bounds(bottom: i32, top: i32, skew_factor: f32, _vertical: bool) -> (f32, f32) {
     let skew_min = ((bottom as f32 + 0.5) * skew_factor).floor();
     let skew_max = ((top as f32 - 0.5) * skew_factor).ceil();
     (skew_min, skew_max)
@@ -237,10 +237,11 @@
     left: i32,
     top: i32,
     skew_factor: f32,
+    vertical: bool, // TODO: vertical skew not yet implemented!
 ) -> (Vec<u8>, usize, i32) {
     let stride = width * 4;
     // Calculate the skewed horizontal offsets of the bottom and top of the glyph.
-    let (skew_min, skew_max) = get_skew_bounds(top - height as i32, top, skew_factor);
+    let (skew_min, skew_max) = get_skew_bounds(top - height as i32, top, skew_factor, vertical);
     // Allocate enough extra width for the min/max skew offsets.
     let skew_width = width + (skew_max - skew_min) as usize;
     let mut skew_buffer = vec![0u8; skew_width * height * 4];
@@ -478,8 +479,12 @@
             if font.flags.contains(FontInstanceFlags::TRANSPOSE) {
                 shape = shape.swap_xy();
             }
+            let (mut tx, mut ty) = (0.0, 0.0);
             if font.synthetic_italics.is_enabled() {
-                shape = shape.synthesize_italics(font.synthetic_italics);
+                let (shape_, (tx_, ty_)) = font.synthesize_italics(shape, y_scale * req_size);
+                shape = shape_;
+                tx = tx_;
+                ty = ty_;
             };
             let mut ft_shape = FT_Matrix {
                 xx: (shape.scale_x * 65536.0) as FT_Fixed,
@@ -487,8 +492,13 @@
                 yx: (shape.skew_y * -65536.0) as FT_Fixed,
                 yy: (shape.scale_y * 65536.0) as FT_Fixed,
             };
+            // The delta vector for FT_Set_Transform is in units of 1/64 pixel.
+            let mut ft_delta = FT_Vector {
+                x: (tx * 64.0) as FT_F26Dot6,
+                y: (ty * -64.0) as FT_F26Dot6,
+            };
             unsafe {
-                FT_Set_Transform(face, &mut ft_shape, ptr::null_mut());
+                FT_Set_Transform(face, &mut ft_shape, &mut ft_delta);
                 FT_Set_Char_Size(
                     face,
                     (req_size * x_scale * 64.0 + 0.5) as FT_F26Dot6,
@@ -660,6 +670,7 @@
                         top - height as i32,
                         top,
                         font.synthetic_italics.to_skew(),
+                        font.flags.contains(FontInstanceFlags::VERTICAL),
                     );
                     left += skew_min as i32;
                     width += (skew_max - skew_min) as i32;
@@ -949,6 +960,7 @@
                         left,
                         top,
                         font.synthetic_italics.to_skew(),
+                        font.flags.contains(FontInstanceFlags::VERTICAL),
                     );
                     final_buffer = skew_buffer;
                     actual_width = skew_width;